File: 30-substring-with-concatenation-of-all-words/substring-with-concatenation-of-all-words.java
class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
        List<Integer> result = new ArrayList<>();
        if (s == null || s.length() == 0 || words == null || words.length == 0) {
            return result;
        }

        int wordLen = words[0].length();
        int wordCount = words.length;
        int totalLen = wordLen * wordCount;
        int sLen = s.length();

        if (sLen < totalLen) return result;

        // Target Map: Recipe parchi
        Map<String, Integer> wordMap = new HashMap<>();
        for (String word : words) {
            wordMap.put(word, wordMap.getOrDefault(word, 0) + 1);
        }

        // wordLen ke shifts par loop chalayenge (e.g., 0, 1, 2 agar wordLen 3 hai)
        for (int i = 0; i < wordLen; i++) {
            int left = i;
            Map<String, Integer> seenWords = new HashMap<>();
            int count = 0;

            // wordLen ke gap par jump lagaye
            for (int j = i; j <= sLen - wordLen; j += wordLen) {
                String currentWord = s.substring(j, j + wordLen);

                if (wordMap.containsKey(currentWord)) {
                    seenWords.put(currentWord, seenWords.getOrDefault(currentWord, 0) + 1);
                    count++;

                    // Duplicate/Extra word aane par left window ko tab tak khisakao jab tak count sahi na ho
                    while (seenWords.get(currentWord) > wordMap.get(currentWord)) {
                        String leftWord = s.substring(left, left + wordLen);
                        seenWords.put(leftWord, seenWords.get(leftWord) - 1);
                        count--;
                        left += wordLen;
                    }

                    // Agar exact saare words match ho gaye
                    if (count == wordCount) {
                        result.add(left);
                    }
                } else {
                    // Invalid word milte hi window clear aur jump to next position
                    seenWords.clear();
                    count = 0;
                    left = j + wordLen;
                }
            }
        }

        return result;
    }
}

File: 30-substring-with-concatenation-of-all-words/substring-with-concatenation-of-all-words.cpp
class Solution {
public:
    vector<int> findSubstring(string s, vector<string>& words) {
          vector<int> result;
        if (s.empty() || words.empty()) return result;

        int wordLen = words[0].length();
        int wordCount = words.size();
        int totalLen = wordLen * wordCount;
        int sLen = s.length();

        if (sLen < totalLen) return result;

        // Target Map: Kaun sa word kitni baar chahiye
        unordered_map<string, int> wordMap;
        for (const string& word : words) {
            wordMap[word]++;
        }

        // Hum string ko wordLen ke alag-alag offsets se scan karenge
        for (int i = 0; i < wordLen; i++) {
            int left = i;
            unordered_map<string, int> seenWords;
            int count = 0; // Kitne valid words mil chuke hain

            // j ko wordLen ki jumps se aage badhayenge
            for (int j = i; j <= sLen - wordLen; j += wordLen) {
                string currentWord = s.substr(j, wordLen);

                // Agar word humari target list me hai
                if (wordMap.find(currentWord) != wordMap.end()) {
                    seenWords[currentWord]++;
                    count++;

                    // Agar koi word limit se zyada baar aa gaya, toh left pointer ko aage badhao
                    while (seenWords[currentWord] > wordMap[currentWord]) {
                        string leftWord = s.substr(left, wordLen);
                        seenWords[leftWord]--;
                        count--;
                        left += wordLen; // Window ko left se chota kiya
                    }

                    // Agar saare words sahi count me mil gaye
                    if (count == wordCount) {
                        result.push_back(left);
                    }
                } 
                // Agar koi aisa word mila jo list me hai hi nahi, toh window ko reset karo
                else {
                    seenWords.clear();
                    count = 0;
                    left = j + wordLen; // Window ab naye sire se shuru hogi
                }
            }
        }

        return result;
    }
};